### PR TITLE
Skiko events update. Text/Key input for (jvm, macos, ios).

### PR DESCRIPTION
**Pull request by Rsedaikin**

### CODE REVIEW

File: skiko/src/linuxMain/kotlin/org/jetbrains/skiko/SkiaLayer.linux.kt [hunk 1]
```diff
@@ -93,6 +93,7 @@ actual enum class SkikoKey(val value: Int) {
     KEY_RIGHT_SHIFT(60),
```
Comment (olonho) [inline skiko/src/linuxMain/kotlin/org/jetbrains/skiko/SkiaLayer.linux.kt:93]: May make sense to move to separate file as well.

File: skiko/src/macosMain/kotlin/org/jetbrains/skiko/SkiaLayer.macos.kt [hunk 2]
```diff
@@ -105,13 +111,60 @@ actual open class SkiaLayer() {
...
             override fun flagsChanged(event: NSEvent) {
                 skikoView?.onKeyboardEvent(toSkikoEvent(event))
             }
             override fun keyUp(event: NSEvent) {
                 skikoView?.onKeyboardEvent(toSkikoEvent(event, SkikoKeyboardEventKind.UP))
             }
+            
+            private var keyEvent: NSEvent? = null
+            private var markedText: String = ""
+            private val kEmptyRange = NSMakeRange(NSNotFound.toULong(), 0.toULong())
```
Comment (olonho) [inline skiko/src/macosMain/kotlin/org/jetbrains/skiko/SkiaLayer.macos.kt:126]: `NSNotFound.convert()` may be better here and in other places.

File: skiko/src/macosMain/kotlin/org/jetbrains/skiko/SkiaLayer.macos.kt [hunk 3]
```diff
@@ -105,13 +111,60 @@ actual open class SkiaLayer() {
...
+            override fun firstRectForCharacterRange(range: CValue<NSRange>, actualRange: NSRangePointer?): CValue<NSRect> {
+                val (x, y) = window.frame.useContents {
+                    origin.x to origin.y
+                }
+                return NSMakeRect(x, y, 0.0, 0.0)
+            }
+            override fun insertText(string: Any, replacementRange: CValue<NSRange>) {
+                var character = ""
+                if (string is NSAttributedString) {
+                    character = string.string
```
Comment (olonho) [inline skiko/src/macosMain/kotlin/org/jetbrains/skiko/SkiaLayer.macos.kt:149]: Hmm, could you please explain this if and cast.

File: skiko/src/linuxMain/kotlin/org/jetbrains/skiko/SkiaLayer.linux.kt (hunk 1 above)
Comment (Rsedaikin) [reply skiko/src/linuxMain/kotlin/org/jetbrains/skiko/SkiaLayer.linux.kt]: Done.

File: skiko/src/iosMain/kotlin/org/jetbrains/skiko/SkikoViewController.kt [hunk 4]
```diff
@@ -10,15 +10,52 @@ import platform.UIKit.UIScreen
...
+    override fun hasText(): Boolean {
+        return inputText.length > 0
+    }
+    override fun insertText(theText: String) {
+        inputText += theText
+        skikoLayer.skikoView?.onInputEvent(toSkikoTypeEvent(theText))
+    }
+    override fun deleteBackward() {
+        inputText = inputText.dropLast(1)
+        skikoLayer.skikoView?.onInputEvent(toSkikoTypeEvent("\b"))
```
[the remark this answers is not in the document]
Comment (Rsedaikin) [reply skiko/src/iosMain/kotlin/org/jetbrains/skiko/SkikoViewController.kt]: This is an extra line of code for my tests - I removed it. Done.

File: skiko/src/macosMain/kotlin/org/jetbrains/skiko/SkiaLayer.macos.kt (hunk 3 above)
Comment (Rsedaikin) [reply skiko/src/macosMain/kotlin/org/jetbrains/skiko/SkiaLayer.macos.kt:149]: It looks like the string could be NSString or NSAttributedString, so you need to check it first.
see apple doc: https://developer.apple.com/documentation/appkit/nstextinputclient/1438258-inserttext

File: skiko/src/macosMain/kotlin/org/jetbrains/skiko/SkiaLayer.macos.kt (hunk 2 above)
Comment (Rsedaikin) [reply skiko/src/macosMain/kotlin/org/jetbrains/skiko/SkiaLayer.macos.kt]: Done.

**Review by olonho — APPROVED** [review]

**Pull request merged**